omegaconf/omegaconf.py

```diff
@@ -124,17 +126,13 @@ def env(key: str, default: Any = _EMPTY_MARKER_, *, config: BaseContainer) -> An
                 return val_str
 
         # Resolve the parse tree.
-        val = config.resolve_parse_tree(parse_tree)
+        assert parent is not None
+        root = OmegaConf.get_root(parent)
+        val = root.resolve_parse_tree(parse_tree)
```

Comment:
This is really not clear to me, what does it have to do with the env resolver?

Reply:
This is to convert the env variable from a string into (maybe) something else, by parsing it with the grammar.

As a reminder, in OmegaConf 2.0, env variables are parsed using custom code for integers, floats and booleans (https://github.com/omry/omegaconf/blob/800e8a7ad30db9400b8332d556520a3ecee56b77/omegaconf/_utils.py#L397). We agreed before that in 2.1 we should instead use the grammar to parse them (which enables more powerful constructs, like lists, dicts and interpolations).

Reply:
I see.
After looking at the whole function (not just the latest changes I understand what you are doing here).

There is a spectrum of options here.
1. Maybe it is better to just return the string as is for the env resolver. Least surprising.
2. Maybe we can parse primitive only and not support interpolations (error out).
3. We can switch to class for resolvers instead of a lambda.
The base class can have a member parent node.
It's a bit more heavyweight, but not too bad.

```python
# base class
calss Resolver:
  parent : Container

class EnvResolver(omegaconf.Resolver):
  def resolve(key) -> Any:
    val_str = os.environ[key]
    ...
    # use self.parent to resolve interpolations

OmegaConf.register_resolver("env", EnvResolver)
```

Any thoughts?
I am leaning toward 1 or 2.

Reply:
I actually like option 1 because of its simplicity. However, doing so would remove the ability to convert an env variable into an int / float / bool, which I believe is a useful feature to keep. We could address this by adding a new resolver (e.g., named `eval`, `parse` or `resolve`) that would take a string as input and evaluate it. But then we're back to the question of providing a way for a resolver to access the config, and if we add such an option then we didn't really gain much in terms of implementation complexity (I do believe it may be easier to understand for users, so I'm ok with going this route if that's your preferred choice, but we'll also need to deprecate the current env resolver so the transition will be a bit more complex).

Option 2: easy to do, though personally I prefer option 4 below since (up to a few cosmetic changes) we have it working already (so why not use it)

Option 3: not a fan, sure it's manageable but it makes things look unnecessarily heavy IMO

Option 4: keep the ability to access the parent node but as an option (so that most users can define simple resolvers without having to care about it)

So my preference would be either 1 or 4 ==> final call is yours

Reply:
Okay, let's go with option 2.
- We can parse primitives (int, bool, float, string, list, dict)
- We can pass interpolations as a string to the function and not resolve them. not sure if this should be an error or not.

I think the general use case of accessing the parent node can be achieved without an API change by passing a relative interpolation like `${.}` as a parameter to the resolver.
Under this assumption, I don't think it's worth making the API more complicated at this time. We can always do it later if we decide (but it's much harder to backtrack).

Reply:
Ok, done in 63132ac

Let me know if I should try to rebase it on top of another commit to make your review easier -- I'm not really sure where you're at in the commit sequence ;)

Regarding using `${.}`, this sounds like a good idea, but FYI at this time my grammar refuses this notation ==> if it's supposed to work now then a test should be added for it in master.

Reply:
I am confused:
> We can pass interpolations as a string to the function and not resolve them. not sure if this should be an error or not.

Reading your example, with plus of revenue, looks like you are still supporting interpolations in environment variables.
Is that limited to custom resolvers? I thought to just reject or ignore interpolations in environment variables.

Reply:
> Reading your example, with plus of revenue, looks like you are still supporting interpolations in environment variables.
> Is that limited to custom resolvers? I thought to just reject or ignore interpolations in environment variables.

Oops, sorry, I meant to say something about this but forgot.

Regarding "not sure if this should be an error or not", I decided to make it an error because (1) it's more explicit, and (2) it would make it easier to add support later if we change our mind.

And yes, I indeed kept the support for resolver interpolations because one doesn't need to access the config for them (it would be more work to make them crash since they work "out of the box").

Reply:
Okay, I think we can keep this undocumented - especially in the context of the env resolver.
the shell also interpret `$` and even `${}` and  it will be super confusing for people to try that in practice.

Reply:
> Okay, I think we can keep this undocumented - especially in the context of the env resolver.
> the shell also interpret `$` and even `${}` and it will be super confusing for people to try that in practice.

Ok, I removed the doc in 30026ae

Reply:
> Ok, done in [63132ac](https://github.com/omry/omegaconf/commit/63132ac16611ae35b97ab49624bae2033d4f7b58)
> 
> Let me know if I should try to rebase it on top of another commit to make your review easier -- I'm not really sure where you're at in the commit sequence ;)
> 
> Regarding using `${.}`, this sounds like a good idea, but FYI at this time my grammar refuses this notation ==> if it's supposed to work now then a test should be added for it in master.

Generally it a recipe for infinite recursion as an ad-hoc value inside an interpolation I think it makes sense.
we can think about adding support for it later.